### PR TITLE
Output a log when GC found too many versions for a single key

### PR DESCRIPTION
**Pull request by breeswish**

### CODE REVIEW

**Comment by breeswish** [comment]
@disksing @BusyJay @zhangjinpeng1987 PTAL

**Comment by MyonKeminta** [comment]
Should this be merged before #3119 ?

**Comment by breeswish** [comment]
@myonkeminta Oh, I guess yes. This will be carried in our next release.

File: src/storage/mvcc/txn.rs [hunk 1]
```diff
@@ -270,7 +270,7 @@ impl MvccTxn {
         Ok(())
     }
 
-    pub fn gc(&mut self, key: &Key, safe_point: u64) -> Result<()> {
+    pub fn gc(&mut self, key: &Key, safe_point: u64) -> Result<(usize, usize)> {
```
Comment (gregwebs) [inline src/storage/mvcc/txn.rs]: The return type will be fairly self-documenting if this switches from a tuple to a struct `GcInfo { versions: usize, delete_versions: usize }`

File: src/storage/mvcc/txn.rs [hunk 2]
```diff
@@ -324,11 +324,11 @@ impl MvccTxn {
             self.delete_write(key, commit);
             delete_versions += 1;
         }
-        MVCC_VERSIONS_HISTOGRAM.observe(f64::from(versions));
+        MVCC_VERSIONS_HISTOGRAM.observe(versions as f64);
```
Comment (gregwebs) [inline src/storage/mvcc/txn.rs]: It is possible to move these observations out of this function now. It makes sense to me to both log and observe in the same place.

File: src/storage/txn/scheduler.rs [hunk 3]
```diff
@@ -67,6 +67,12 @@ pub const GC_BATCH_SIZE: usize = 512;
 // The write batch will be around 32KB if we scan 256 keys each time.
 pub const RESOLVE_LOCK_BATCH_SIZE: usize = 256;
 
+/// After scanned how many versions for a key we should output a log for it.
```
Comment (gregwebs) [inline src/storage/txn/scheduler.rs]: After the GC scan of a key, output a message to the log if there are at least this many versions of the key.

File: src/storage/txn/scheduler.rs [hunk 4]
```diff
@@ -67,6 +67,12 @@ pub const GC_BATCH_SIZE: usize = 512;
...
+const GC_LOG_FOUND_VERSION_THRESHOLD: usize = 300;
```
Comment (siddontang) [inline src/storage/txn/scheduler.rs]: I think different GC interval should have different version threshold, e.g. for 10 m, maybe 10 or 20 version is too large, but for 1 d, maybe 300.

Comment (breeswish) [reply src/storage/txn/scheduler.rs]: Good idea!

Comment (breeswish) [reply src/storage/txn/scheduler.rs]: It seems that we need to change the protocol, adding something like "last gc", to support the interval. Any better solutions?

Comment (siddontang) [reply src/storage/txn/scheduler.rs]: I have no idea now. Maybe we should use a smaller value, 300 is too big, I guess 20 or 30 is fine now.

File: src/storage/txn/scheduler.rs [hunk 5]
```diff
@@ -862,7 +870,29 @@ fn process_write_impl(
...
+                        "[region {}] GC found at least {} versions for key {}",
+                        ctx.get_region_id(),
+                        gc_info.found_versions,
+                        k
+                    );
+                }
+                // We may delete only part of the versions a time, which may not beyond the logging
+                // threshold `GC_LOG_DELETED_VERSION_THRESHOLD`.
+                if gc_info.deleted_versions as usize >= GC_LOG_DELETED_VERSION_THRESHOLD
+                    || txn.write_size() >= MAX_TXN_WRITE_SIZE
```
Comment (zhangjinpeng1987) [inline src/storage/txn/scheduler.rs]: I don't think the condition `txn.write_size() >= MAX_TXN_WRITE_SIZE` is ok here.

File: src/storage/txn/scheduler.rs (hunk 4 above)
Comment (breeswish) [reply src/storage/txn/scheduler.rs]: Ok. @zhangjinpeng1987 do you have suggestions?

**Comment by breeswish** [comment]
Updated, @gregwebs PTAL again, thanks! The metric observations are not moved, since it updates related metrics belongs to the MVCC package, and it will be wired to only move some of them to the outside. You can consider the logging as a higher-level work, where there is a `region_id` concept but inside MVCC there isn't one.

File: src/storage/txn/scheduler.rs (hunk 5 above)
Comment (breeswish) [reply src/storage/txn/scheduler.rs]: That's ok for current threshold. But in future, if the threshold is large and we cannot reach the threshold in a single batch (txn.write_size() >= MAX_TXN_WRITE_SIZE) I think we still need to log an item here. I will change it to a comment for reminders.

**Pull request merged**